File: facade-pattern/facade_pattern.rb
=begin
What is the Facade Pattern?

- When you create a simplified interface that performs many
  other actions behind the scenes

Ex:
  - Can I withdrawal $50 from the bank?
  - Check if the checking account is valid
  - Chceck if security code is valid
  - Chceck if fund are available
  - Make changes accordingly

@see http://www.newthinktank.com/2012/09/facade-design-pattern-tutorial/
=end

class WelcomeBank
  def initialize
    puts 'Welcome to ABC Bank'
  end
end

class AccountNumberCheck
  def account_active?(account_number)
    account_number == 12345678
  end
end

class SecurityCodeCheck
  def correct_code?(security_code)
    security_code == 1234
  end
end

class FundsCheck
  def initialize
    @cash_in_account = 100000
  end

  def decrease_cash_in_account(cash_withdraw)
    @cash_in_account = @cash_in_account - cash_withdraw
  end

  def increase_cash_in_account(cash_deposit)
    @cash_in_account = @cash_in_account + cash_deposit
  end

  def enough_money?(cash_to_withdraw)
    if cash_to_withdraw > @cash_in_account
      puts "[error] you don't have enough money"
      puts "[error] current balance: #{@cash_in_account}"
      false
    else
      decrease_cash_in_account(cash_to_withdraw)
      puts "[withdraw complete] current balance is: #{@cash_in_account}"
      true
    end
  end

  def make_deposit(cash_to_deposit)
    increase_cash_in_account(cash_to_deposit)
    puts "[deposit complete] current balance is: #{@cash_in_account}"
  end
end

class BankAccountFacade
  attr_reader :account_number,
              :security_code,
              :welcome_bank,
              :account_checker,
              :code_checker,
              :fund_checker

  def initialize(account_number, security_code)
    @account_number = account_number
    @security_code = security_code
    @welcome_bank = WelcomeBank.new
    @account_checker = AccountNumberCheck.new
    @code_checker = SecurityCodeCheck.new
    @fund_checker = FundsCheck.new
  end

  def withdraw_cash(cash)
    if account_checker.account_active?(account_number) &&
       code_checker.correct_code?(security_code) &&
       fund_checker.enough_money?(cash)
      puts 'Transaction complete'
    else
      puts 'Transaction failed'
    end
  end

  def deposit_cash(cash)
    if account_checker.account_active?(account_number) &&
       code_checker.correct_code?(security_code)
      fund_checker.make_deposit(cash)
      puts 'Transaction complete'
    else
      puts 'Transaction failed'
    end
  end
end

################################################################################

accessing_bank = BankAccountFacade.new(12345678, 1234)
accessing_bank.withdraw_cash(5000)
accessing_bank.withdraw_cash(90000)
accessing_bank.deposit_cash(20000)

File: facade-pattern/facade-pattern.ts
/**
 * What is the Facade Pattern?
 *
 * - When you create a simplified interface that performs many
 *   other actions behind the scenes
 *
 * Ex:
 *   - Can I withdrawal $50 from the bank?
 *     - Check if the checking account is valid
 *     - Chceck if security code is valid
 *     - Chceck if fund are available
 *     - Make changes accordingly
 *
 * @see http://www.newthinktank.com/2012/09/facade-design-pattern-tutorial/
 */

class WelcomeToBank {
  public constructor() {
    console.log("Welcome to ABC Bank");
  }
}

class AccountNumberCheck {
  readonly accountNumber: number = 12345678;

  public accountActive(accountNumber: number): boolean {
    return accountNumber == this.accountNumber;
  }
}

class SecurityCodeCheck {
  readonly securityCode: number = 1234;

  public isCodeCorrect(securityCode: number): boolean {
    return securityCode == this.securityCode;
  }
}

class FundsCheck {
  #cashInAccount: number = 1000;

  get cashInAccount(): number {
    return this.#cashInAccount;
  }

  public decreaseCashInAccount(cashWithDraw: number): void {
    this.#cashInAccount -= cashWithDraw;
  }

  public increaseCashInAccount(cashDeposited: number): void {
    this.#cashInAccount += cashDeposited;
  }

  public haveEnoughMoney(cashToWithdrawal: number): boolean {
    return this.#cashInAccount >= cashToWithdrawal;
  }

  public makeDeposit(cashToDeposit: number): void {
    this.increaseCashInAccount(cashToDeposit);
  }
}

class BankAccountFacade {
  private readonly accChecker: AccountNumberCheck;
  private readonly codeChecker: SecurityCodeCheck;
  private readonly fundChecker: FundsCheck;
  private readonly bankWelcome: WelcomeToBank;

  public constructor(
    readonly accountNumber: number,
    readonly securityCode: number
  ) {
    this.bankWelcome = new WelcomeToBank();
    this.accChecker = new AccountNumberCheck();
    this.codeChecker = new SecurityCodeCheck();
    this.fundChecker = new FundsCheck();
  }

  public withdrawCash(cashToGet: number): void {
    if (
      this.accChecker.accountActive(this.accountNumber) &&
      this.codeChecker.isCodeCorrect(this.securityCode) &&
      this.fundChecker.haveEnoughMoney(cashToGet)
    ) {
      this.fundChecker.decreaseCashInAccount(cashToGet);
      console.log("transaction complete");
    } else {
      console.log("unable to complete transaction");
    }
  }

  public depositCash(cashToDeposit: number): void {
    if (
      this.accChecker.accountActive(this.accountNumber) &&
      this.codeChecker.isCodeCorrect(this.securityCode)
    ) {
      this.fundChecker.makeDeposit(cashToDeposit);
      console.log("transaction complete");
    } else {
      console.log("unable to complete transaction");
    }
  }
}

// ----------------------------------------------------------------------------

const accessingBank = new BankAccountFacade(12345678, 1234);
accessingBank.withdrawCash(500);
accessingBank.withdrawCash(501);
accessingBank.depositCash(1);
accessingBank.withdrawCash(501);
